src/sage/rings/lazy_series.py

```diff
@@ -1076,35 +1099,151 @@ def __bool__(self):
-        if isinstance(self._coeff_stream, Stream_exact):
+
+        prec = self.parent().options['halting_precision']
+        if prec is None and not self.parent().options['secure']:
             return True
-        if isinstance(self._coeff_stream, Stream_uninitialized):
-            if self._coeff_stream._target is None:
-                return True
-            if isinstance(self._coeff_stream._target, Stream_zero):
-                return False
-            if isinstance(self._coeff_stream._target, Stream_exact):
-                return True
-        if self._coeff_stream._is_sparse:
-            cache = self._coeff_stream._cache
-            if any(cache[a] for a in cache):
-                return True
-        else:
-            if any(self._coeff_stream._cache):
-                return True
 
-        v = self._coeff_stream._approximate_order
-        if self[v]:
+        if isinstance(self._coeff_stream, Stream_exact):
+            return True
+        if self._coeff_stream.is_uninitialized():
+            return True
+        if self._coeff_stream.is_nonzero():
             return True
 
-        prec = self.parent().options['halting_precision']
         if prec is None:
-            raise ValueError("undecidable as lazy Laurent series")
+            raise ValueError("undecidable")
+        v = self._coeff_stream._approximate_order
         return any(self[i] for i in range(v, v + prec))
 
+    def is_nonzero(self):
+        r"""
+        Return ``True`` if ``self`` is *known* to be nonzero.
+
+        EXAMPLES:
+
+        A series that it not known to be nonzero with no halting precision::
+
+            sage: L.<z> = LazyLaurentSeriesRing(GF(2))
+            sage: f = L(lambda n: 0, valuation=0)
+            sage: f.is_nonzero()
+            False
+            sage: bool(f)
+            True
+            sage: g = L(lambda n: 0 if n < 50 else 1, valuation=2)
+            sage: g.is_nonzero()
+            False
+            sage: g[60]
+            1
+            sage: g.is_nonzero()
+            True
+
+        With finite halting precision, it can be considered to
+        be indistinguishable from zero until possibly enough
+        coefficients are computed::
+
+            sage: L.options.halting_precision = 20
+            sage: f = L(lambda n: 0, valuation=0)
+            sage: f.is_zero()
+            True
+
+            sage: g = L(lambda n: 0 if n < 50 else 1, valuation=2)
+            sage: g.is_nonzero()  # checks up to degree 22 = 2 + 20
+            False
+            sage: g.is_nonzero()  # checks up to degree 42 = 22 + 20
+            False
+            sage: g.is_nonzero()  # checks up to degree 62 = 42 + 20
+            True
+            sage: L.options._reset()
+        """
+        if self._coeff_stream.is_nonzero():
+            return True
+        if self.parent().options['halting_precision'] is not None:
+            return bool(self)
+        return False
+
+    def prove_nonzero(self):
+        r"""
```

Comment:
Maybe we could merge this with `is_nonzero` using a keyword argument `proof=True`?

I find the method name extremely ugly.

Reply:
That is a much better idea. I have changed this.